test/support/utils.rb

```diff
@@ -12,10 +12,51 @@ def type(command)
     end
 
     def create_message fail_msg, test_info
-      "#{fail_msg} on #{test_info.mode} mode\n[DEBUGGER SESSION LOG]\n> #{test_info.backlog.join('> ')}#{debuggee_backlog test_info}"
+      debugger_msg = <<~DEBUGGER_MSG.chomp
+        ======================
+        ▼  Debugger Session  ▼
+        ======================
+
+        > #{test_info.backlog.join('> ')}
+
+        ======================
+        ▲  Debugger Session  ▲
+        ======================
+      DEBUGGER_MSG
+
+      debuggee_backlog = collect_debuggee_backlog(test_info)
+      debuggee_msg =
+        if debuggee_backlog && !debuggee_backlog.empty?
```

Comment:
@ko1 @ono-max should I always show debuggee backlog even if they're empty and only hide them in the local mode?
I can't reproduce a failed remote test that contains debuggee backlog locally so I'm not sure what you'd like to see in this case.

Reply:
> should I always show debuggee backlog even if they're empty and only hide them in the local mode?

There is no such things that they're empty. When debuggee is started, the logs will be output as follows.
```shell
Failure: test_conditional_breakpoint_stops_if_condition_is_true(DEBUGGER__::BreakAtLinesTest):
  #<Test::Unit::AssertionFailedError: TIMEOUT ERROR (10 sec) on UNIX DOMAIN SOCKET mode
  [DEBUGGER SESSION LOG]
  > [1, 10] in /var/folders/kv/w1k6nh1x5fl7vx47b2pd005w0000gn/T/debugger20210829-7101-9abjlb.rb
  > =>   1| module Foo
  >      2|   class Bar
  >      3|     def self.a
  >      4|       "hello"
  >      5|     end
  >      6|
  >      7|     def b(n)
  >      8|       2.times do
  >      9|         n
  >     10|       end
  > =>#0	<main> at /var/folders/kv/w1k6nh1x5fl7vx47b2pd005w0000gn/T/debugger20210829-7101-9abjlb.rb:1
  > break if: n == 1
  >
  > (rdbg:remote) break if: n == 1
#0  BP - Check  n == 1
  >
  > (rdbg:remote) continue

  [DEBUGGEE SESSION LOG]
  > EBUGGER: Session start (pid: 7227)
  > DEBUGGER: Debugger can attach via UNIX domain socket (/Users/naotto/.ruby-debug-sock/ruby-debug-naotto-7101)
  > DEBUGGER: wait for debugger connection...
  > DEBUGGER: Connected.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Baz:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  .>.
/Users/naotto/workspace/debug/test/support/assertions.rb:66:in `assert_block'
/Users/naotto/workspace/debug/test/support/utils.rb:64:in `rescue in debug_code'
/Users/naotto/workspace/debug/test/support/utils.rb:52:in `debug_code'
test/debug/break_test.rb:471:in `test_conditional_breakpoint_stops_if_condition_is_true'
     468:     end
     469:
     470:     def test_conditional_breakpoint_stops_if_condition_is_true
  => 471:       debug_code(program) do
     472:         type 'break if: n == 1'
     473:         assert_line_text(/#0  BP - Check  n == 1/)
     474:         type 'continue'
```

Reply:
Not on my machine (mac)

```
Failure: test_tracer_prints_correct_method_receiving_messages(DEBUGGER__::TracePassTest::TraceCallReceiverTest):
  TIMEOUT ERROR (10 sec) on UNIX DOMAIN SOCKET mode
  [DEBUGGER SESSION LOG]
  > [1, 10] in /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb
  > =>   1| class Foo
  >      2|   def bar; end
  >      3|   def self.baz; end
  >      4| end
  >      5|
  >      6| class Bar < Foo
  >      7| end
  >      8|
  >      9| f = Foo.new
  >     10| b = Bar.new
  > =>#0        <main> at /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb:1
  > c
  >
  > (rdbg:remote) c
[10, 19] in /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb
  >     10| b = Bar.new
  >     11|
  >     12| def f.foobar; end
  >     13| def b.foobar; end
  >     14|
  > =>  15| binding.b
  >     16|
  >     17|  Foo.baz
  >     18|  f.bar
  >     19|  f.foobar
  > =>#0        <main> at /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb:15
  >
  > (rdbg:remote) trace pass Foo
Enable PassTracer for Foo (enabled)
  >
  > (rdbg:remote) trace pass fq
eval error: undefined local variable or method `fq' for main:Object
  > Did you mean?  f
  >   (rdbg)//var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb:1:in `<main>'
  > undefined local variable or method `fq' for main:Object
  > Did you mean?  f
  >
  > (rdbg:remote) c
DEBUGGER (trace/pass) #th:1 #depth:2 `Foo` receives .baz (#<Class:Foo>.baz) at /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb:3
  > [20, 25] in /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb
  >     20|
  >     21|  Bar.baz
  >     22|  b.bar
  >     23|  b.foobar
  >     24|
  > =>  25|  binding.b
  > =>#0        <main> at /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-43632-12yszap.rb:25

  [DEBUGGEE SESSION LOG]
  >
/Users/st0012/projects/debug/test/support/assertions.rb:66:in `assert_block'
/Users/st0012/projects/debug/test/support/utils.rb:193:in `rescue in block in run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:189:in `block in run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:140:in `spawn'
/Users/st0012/projects/debug/test/support/utils.rb:140:in `run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:126:in `debug_on_unix_domain_socket'
/Users/st0012/projects/debug/test/support/utils.rb:68:in `debug_code'
test/debug/trace_test.rb:317:in `test_tracer_prints_correct_method_receiving_messages'
     314:       end
     315:
     316:       def test_tracer_prints_correct_method_receiving_messages
  => 317:         debug_code(program) do
     318:           type 'c'
     319:           type 'trace pass Foo'
     320:           type 'trace pass fq'
```

Reply:
Hmm, my machine is also mac though...

Reply:
I thought you don’t use macOS?

Reply:
I use macOS lol

Reply:
How come we see different behaviors on errno::EIO though 🤔

Reply:
I'm not sure what's going on here, but I know that `errno::EIO` occurs on ubuntu-latest(Github Workflow). Your environment may be similar to it?

Reply:
ah I thought you added it because you can see it in your environment 🤦 
anyway, I think I didn't see the debuggee session because of the test case. I used the same one as you used above `test_conditional_breakpoint_stops_if_condition_is_true ` and now I can see it:

```
Failure: test_conditional_breakpoint_stops_if_condition_is_true(DEBUGGER__::BreakAtLinesTest):

  ======================
  ▼  Debugger Session  ▼
  ======================

  > [1, 10] in /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-45541-1lt4unl.rb
  > =>   1| module Foo
  >      2|   class Bar
  >      3|     def self.a
  >      4|       "hello"
  >      5|     end
  >      6|
  >      7|     def b(n)
  >      8|       2.times do
  >      9|         n
  >     10|       end
  > =>#0        <main> at /var/folders/yg/hnbymwxd5pn7v94_clc59y6r0000gn/T/debugger20210829-45541-1lt4unl.rb:1
  > break if: n == 1
  >
  > (rdbg:remote) break if: n == 1
#0  BP - Check  n == 1
  >
  > (rdbg:remote) continue


  ======================
  ▲  Debugger Session  ▲
  ======================
  ======================
  ▼  Deguggee Session  ▼
  ======================

  > EBUGGER: Session start (pid: 45555)
  > DEBUGGER: Debugger can attach via UNIX domain socket (/Users/st0012/.ruby-debug-sock/ruby-debug-st0012-45541)
  > DEBUGGER: wait for debugger connection...
  > DEBUGGER: Connected.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Baz:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo::Bar:Class (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.
  > [EVAL ERROR]
  >   expr: n == 1
  >   err: undefined local variable or method `n' for Foo:Module (NameError)
  > Error caused by  BP - Check  n == 1.


  ======================
  ▲  Debuggee Session  ▲
  ======================
  ===================
  ▼ Failure Message ▼
  ===================

  TIMEOUT ERROR (10 sec) on UNIX DOMAIN SOCKET mode
/Users/st0012/projects/debug/test/support/assertions.rb:66:in `assert_block'
/Users/st0012/projects/debug/test/support/utils.rb:231:in `rescue in block in run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:227:in `block in run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:181:in `spawn'
/Users/st0012/projects/debug/test/support/utils.rb:181:in `run_test_scenario'
/Users/st0012/projects/debug/test/support/utils.rb:167:in `debug_on_unix_domain_socket'
/Users/st0012/projects/debug/test/support/utils.rb:109:in `debug_code'
test/debug/break_test.rb:471:in `test_conditional_breakpoint_stops_if_condition_is_true'
     468:     end
     469:
     470:     def test_conditional_breakpoint_stops_if_condition_is_true
  => 471:       debug_code(program) do
     472:         type 'break if: n == 1'
     473:         assert_line_text(/#0  BP - Check  n == 1/)
     474:         type 'continue'
```